src/channels/mpsc.go

```diff
@@ -0,0 +1,59 @@
+package channels
+
+// Multiple Producer Single Consumer channel (mpsc channel)
+type MPSC[T any] struct {
+	input  []<-chan T
+	output chan<- T
+}
+
+func NewMPSC[T any](output chan<- T, blocking bool) *MPSC[T] {
+	channel := &MPSC[T]{
+		input:  make([]<-chan T, 0),
+		output: output,
+	}
+
+	if blocking {
+		go channel.runBlocking()
+	} else {
+		go channel.run()
+	}
+
+	return channel
+}
+
+func (mpsc *MPSC[T]) run() {
+	for {
+		for _, channel := range mpsc.input {
+			select {
+			case payload := <-channel:
+				select {
+				case mpsc.output <- payload:
+				default:
+				}
+			default:
+			}
+		}
+	}
+}
+
+func (mpsc *MPSC[T]) runBlocking() {
+	for {
+		for _, channel := range mpsc.input {
+			select {
+			case payload := <-channel:
+				mpsc.output <- payload
+			default:
+			}
+		}
+	}
+}
+
+func (mpsc *MPSC[T]) Add(input <-chan T) {
```

Comment:
You may need a mutex to avoid concurrent iteration and modification of slice

Reply:
That doesnt happen because you are appending to the slice. When iterating in a for loop using range, range is evaluated before iterating, thus, this change doesn't take effect until the next iteration without causing any problems.